### ISSUE TITLE
Typescript support

### ISSUE THREAD

**Issue by yytsui** [appbaseio/reactivesearch#223]
**Issue Type:enhancement**

**Description:**

Any plan to support  [type definition file](https://github.com/DefinitelyTyped/DefinitelyTyped) for Typescript?
**Browser:** [all | Chrome XX | Firefox XX | Edge XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

**Anything else:**

Thanks!

**Comment by romainquellec** [comment]
> 2.3.0, someday

It's time ! (Will see how to do it)

**Comment by metagrover** [comment]
That'd be awesome. We have been swamped with the stuff on `reactivemaps` among other things 😅 

Feel free to reach out to me here or on [gitter](https://gitter.im/appbaseio/reactivesearch) if you need any help or suggestions. Looking forward to see the typescript support 😄

**Comment by romainquellec** [comment]
@yytsui, do you know how to make it ?

**Comment by romainquellec** [comment]
In short, we need something like `import * as React from 'react';

declare namespace DataController {
    export interface DataControllerProps {
      addComponent: void,
    	removeComponent: void,
    	updateQuery: void,
    	selectedValue?: any,
    	componentId: string,
    	beforeValueChange?: void,
    	children?: any,
    	className?: string,
    	customQuery?: void,
    	defaultSelected?: any,
    	filterLabel?: string,
    	onQueryChange?: void,
    	onValueChange?: void,
    	showFilter?: boolean,
    	style?: any,
    	URLParams: boolean,
    }
    //export interface DataControllerState {}
}
//declare class DataController extends React.Component<DataController.DataControllerProps, DataController.DataControllerState> { }
declare class DataController extends React.Component<DataController.DataControllerProps> { }
export default DataController;`

For each components.

**Comment by metagrover** [comment]
Yeah, that's pretty much it. I agree it is quite a bit of work 😅 

We have been looking into [material-ui](https://github.com/mui-org/material-ui/tree/v1-beta/src/Tabs)'s type definitions and we will probably use a similar pattern. 

Here's how I'm thinking of achieving this: Start with defining a standard component type definitions (should be simple, since most of the components share same method naming) and then extending that definition to create typings for every component.

**Comment by romainquellec** [comment]
Well, I don't have so much knowledge about your components. How can we work on this ?

**I'm proposing this :** 
You (or your team) create a new repo on appbaseio as a working version, then we merge it to DefinitelyTyped ?
So, I can submit some code on this and that. Are you ok with that ?

Thanks !

**Comment by romainquellec** [comment]
There is not much now, but you can use [this](https://github.com/romainquellec/appbaseio__reactivesearch)

**Comment by romainquellec** [comment]
Maybe it can help : https://yarnpkg.com/en/package/react-to-typescript-definitions

**Comment by metagrover** [comment]
Hey @romainquellec! Thanks, this looks interesting. I will try it over the weekend (Sorry, I got busy with other things and this issue skipped my mind)

In any case, I will set up a separate branch here with the standard component typedefs in place in a day or two, and then we can take it further from there.

Will keep the progress posted here!

**Comment by metagrover** [comment]
I tried `react-to-typescript-definitions` and I was able to generate the type defs for ReactiveBase component as follows:

```
declare module 'ReactiveBase' {
    import * as React from 'react';

    export interface ReactiveBaseProps {
        app?: any;
        children?: any;
        credentials?: any;
        headers?: any;
        queryParams?: any;
        theme?: any;
        themePreset?: any;
        type?: any;
        url?: any;
        mapKey?: any;
        style?: any;
        className?: any;
    }

    export default class ReactiveBase extends React.Component<ReactiveBaseProps, any> {
        render(): JSX.Element;
    }
}
```

via

```
cat ./path/to/ReactiveBase.js | npx react2dts --module-name "ReactiveBase"
```

Notice how every prop is of type `any`. And, for every other component, I got:

```
declare module 'TextField' {
    import * as React from 'react';

}
```
as the output. Not sure why. 

Note that all the components except `ReactiveBase`, are exposed via `connect()` method from redux which could be the cause. I also tried some other solutions, but no luck yet 😞.

**Comment by metagrover** [comment]
Primary support has been added. Check [typescript](https://github.com/appbaseio/reactivesearch/tree/typescript) branch for details. After some tests, we will roll this out in the upcoming release.

All credits to @dhruvdutt 🎉
